File: sandbox/src/streaming/exec-stream.ts
import type { ExecResult, ExecStreamFrame } from '../types';
import { parseSseFrames } from './sse';

export type ExecLog = {
  stream: 'stdout' | 'stderr';
  data: string;
};

export type ExecStreamOptions = {
  signal?: AbortSignal;
};

function aggregateFrames(frames: ExecStreamFrame[]): ExecResult {
  let stdout = '';
  let stderr = '';
  let exit_code = 1;
  let duration_ms = 0;
  let sawDone = false;

  for (const frame of frames) {
    if (frame.type === 'stdout') {
      stdout += frame.data;
    }

    if (frame.type === 'stderr') {
      stderr += frame.data;
    }

    if (frame.type === 'done') {
      exit_code = frame.exit_code;
      duration_ms = frame.duration_ms;
      sawDone = true;
    }

    if (frame.type === 'error') {
      throw new Error(frame.message);
    }
  }

  if (!sawDone) {
    throw new Error('Command stream ended before completion');
  }

  return { stdout, stderr, exit_code, duration_ms };
}

/**
 * Live command output from a streaming exec or runCode call.
 *
 * ```
 * const output = await sandbox.exec({ cmd: 'npm install', stream: true });
 * for await (const log of output) {
 *   if (log.stream === 'stdout') process.stdout.write(log.data);
 * }
 * const result = await output.result();
 * ```
 */
export class ExecStream implements AsyncIterable<ExecLog> {
  private readonly body: ReadableStream<Uint8Array>;
  private cachedFrames: ExecStreamFrame[] | null = null;
  private consumePromise: Promise<ExecStreamFrame[]> | null = null;

  public constructor(body: ReadableStream<Uint8Array>) {
    this.body = body;
  }

  /**
   * Iterate stdout/stderr chunks as they arrive.
   * Stops before the terminal `done` frame.
   */
  public async *logs(options: ExecStreamOptions = {}): AsyncGenerator<ExecLog> {
    for await (const frame of this.frames(options)) {
      if (frame.type === 'stdout') {
        yield { stream: 'stdout', data: frame.data };
      }

      if (frame.type === 'stderr') {
        yield { stream: 'stderr', data: frame.data };
      }
    }
  }

  /** Iterate all SSE frames, including the terminal `done` frame. */
  public async *frames(options: ExecStreamOptions = {}): AsyncGenerator<ExecStreamFrame> {
    if (this.cachedFrames) {
      for (const frame of this.cachedFrames) {
        yield frame;
      }
      return;
    }

    const frames: ExecStreamFrame[] = [];

    for await (const frame of parseSseFrames(this.body, options.signal)) {
      frames.push(frame);
      yield frame;
    }

    this.cachedFrames = frames;
  }

  /** Wait for completion and return the aggregated command result. */
  public async result(options: ExecStreamOptions = {}): Promise<ExecResult> {
    if (this.cachedFrames) {
      return aggregateFrames(this.cachedFrames);
    }

    if (!this.consumePromise) {
      this.consumePromise = this.collectFrames(options);
    }

    const frames = await this.consumePromise;
    return aggregateFrames(frames);
  }

  public [Symbol.asyncIterator](): AsyncIterator<ExecLog> {
    return this.logs()[Symbol.asyncIterator]();
  }

  private async collectFrames(options: ExecStreamOptions = {}): Promise<ExecStreamFrame[]> {
    const frames: ExecStreamFrame[] = [];

    for await (const frame of this.frames(options)) {
      frames.push(frame);
    }

    return frames;
  }
}

export async function consumeExecStream(
  stream: ExecStream,
  callbacks: {
    onStdout?: (chunk: string) => void;
    onStderr?: (chunk: string) => void;
  } = {},
  options: ExecStreamOptions = {},
): Promise<ExecResult> {
  for await (const frame of stream.frames(options)) {
    if (frame.type === 'stdout') {
      callbacks.onStdout?.(frame.data);
    }

    if (frame.type === 'stderr') {
      callbacks.onStderr?.(frame.data);
    }

    if (frame.type === 'error') {
      throw new Error(frame.message);
    }
  }

  return stream.result(options);
}

File: sandbox-python/brimble_sandbox/streaming/exec_stream.py
"""Parsed exec/code streaming output."""

from __future__ import annotations

from collections.abc import Callable, Iterator
from typing import TYPE_CHECKING, cast

from ..types import ExecLog, ExecResult, ExecStreamFrame
from .sse import parse_sse_frames

if TYPE_CHECKING:
    import requests


def _aggregate_frames(frames: list[ExecStreamFrame]) -> ExecResult:
    stdout = ""
    stderr = ""
    exit_code = 1
    duration_ms = 0
    saw_done = False

    for frame in frames:
        frame_type = frame["type"]
        if frame_type == "stdout":
            stdout += frame["data"]
        elif frame_type == "stderr":
            stderr += frame["data"]
        elif frame_type == "done":
            exit_code = frame["exit_code"]
            duration_ms = frame["duration_ms"]
            saw_done = True
        elif frame_type == "error":
            raise RuntimeError(frame["message"])

    if not saw_done:
        raise RuntimeError("Command stream ended before completion")

    return {
        "stdout": stdout,
        "stderr": stderr,
        "exit_code": exit_code,
        "duration_ms": duration_ms,
    }


class ExecStream:
    """
    Live command output from a streaming exec or run_code call.

    ```python
    output = sandbox.exec_stream({"cmd": "npm install"})
    for log in output:
        if log["stream"] == "stdout":
            print(log["data"], end="")
    result = output.result()
    ```
    """

    def __init__(self, response: requests.Response) -> None:
        self._response = response
        self._cached_frames: list[ExecStreamFrame] | None = None

    def logs(self) -> Iterator[ExecLog]:
        """Iterate stdout/stderr chunks as they arrive."""
        for frame in self.frames():
            frame_type = frame["type"]
            if frame_type == "stdout":
                yield {"stream": "stdout", "data": frame["data"]}
            elif frame_type == "stderr":
                yield {"stream": "stderr", "data": frame["data"]}

    def frames(self) -> Iterator[ExecStreamFrame]:
        """Iterate all SSE frames, including the terminal `done` frame."""
        if self._cached_frames is not None:
            yield from self._cached_frames
            return

        frames: list[ExecStreamFrame] = []
        body = self._response.raw
        if body is None:
            raise RuntimeError("Expected a response stream but received an empty body")

        for frame in parse_sse_frames(body):
            frames.append(frame)
            yield frame

        self._cached_frames = frames

    def result(self) -> ExecResult:
        """Wait for completion and return the aggregated command result."""
        if self._cached_frames is not None:
            return _aggregate_frames(self._cached_frames)

        frames = list(self.frames())
        return _aggregate_frames(frames)

    def close(self) -> None:
        self._response.close()

    def __iter__(self) -> Iterator[ExecLog]:
        return self.logs()

    def __enter__(self) -> ExecStream:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()


def consume_exec_stream(
    stream: ExecStream,
    *,
    on_stdout: Callable[[str], None] | None = None,
    on_stderr: Callable[[str], None] | None = None,
) -> ExecResult:
    """Drain a stream, invoking callbacks for live output."""
    for frame in stream.frames():
        frame_type = frame["type"]
        if frame_type == "stdout":
            if on_stdout is not None:
                on_stdout(frame["data"])
        elif frame_type == "stderr":
            if on_stderr is not None:
                on_stderr(frame["data"])
        elif frame_type == "error":
            raise RuntimeError(cast(dict[str, str], frame)["message"])

    return stream.result()
